File: web/src/__generated/services/LoginTransferController.ts
import type {Executor} from '../';
import type {LoginTransferDto} from '../model/dto/';
import type {
    LoginTransferClaimRequest, 
    LoginTransferCreateResponse, 
    LoginTransferUpdateRequest, 
    LoginTransferUpdateResponse, 
    TokenLoginResponse
} from '../model/static/';

export class LoginTransferController {
    
    constructor(private executor: Executor) {}
    
    readonly cancel: (options: LoginTransferControllerOptions['cancel']) => Promise<
        void
    > = async(options) => {
        let _uri = '/api/login-transfers/';
        _uri += encodeURIComponent(options.id);
        return (await this.executor({uri: _uri, method: 'DELETE'})) as Promise<void>;
    }
    
    /**
     * 新设备认领登录交接
     * 
     * 二维码密钥同时是查询索引与凭据，因此不需要路径参数；成功后一次性下发认领访问令牌。
     */
    readonly claim: (options: LoginTransferControllerOptions['claim']) => Promise<
        LoginTransferUpdateResponse
    > = async(options) => {
        let _uri = '/api/login-transfers/claims';
        return (await this.executor({uri: _uri, method: 'POST', body: options.body})) as Promise<LoginTransferUpdateResponse>;
    }
    
    readonly create: () => Promise<
        LoginTransferCreateResponse
    > = async() => {
        let _uri = '/api/login-transfers';
        return (await this.executor({uri: _uri, method: 'POST'})) as Promise<LoginTransferCreateResponse>;
    }
    
    readonly createToken: (options: LoginTransferControllerOptions['createToken']) => Promise<
        TokenLoginResponse
    > = async(options) => {
        let _uri = '/api/login-transfers/';
        _uri += encodeURIComponent(options.id);
        _uri += '/tokens';
        const _headers: {[key:string]: string} = {};
        if (options.authorization) {
            _headers['Authorization'] = options.authorization
        }
        return (await this.executor({uri: _uri, method: 'POST', headers: _headers})) as Promise<TokenLoginResponse>;
    }
    
    /**
     * 查询登录交接状态
     * 
     * 原设备凭账号会话查询，新设备凭认领访问令牌查询，后者只获得完成流程所需的最小状态。
     * 已进入终态（含过期）的交接同样返回 `200`，由调用方读取 `status` 判断。
     */
    readonly get: (options: LoginTransferControllerOptions['get']) => Promise<
        LoginTransferDto['LoginTransferController/SOURCE_TRANSFER_FETCHER']
    > = async(options) => {
        let _uri = '/api/login-transfers/';
        _uri += encodeURIComponent(options.id);
        const _headers: {[key:string]: string} = {};
        if (options.authorization) {
            _headers['Authorization'] = options.authorization
        }
        return (await this.executor({uri: _uri, method: 'GET', headers: _headers})) as Promise<LoginTransferDto['LoginTransferController/SOURCE_TRANSFER_FETCHER']>;
    }
    
    /**
     * 原设备审批：把已认领的交接置为 AUTHORIZED 或 REJECTED。
     */
    readonly update: (options: LoginTransferControllerOptions['update']) => Promise<
        LoginTransferUpdateResponse
    > = async(options) => {
        let _uri = '/api/login-transfers/';
        _uri += encodeURIComponent(options.id);
        const _headers: {[key:string]: string} = {};
        if (options.authorization) {
            _headers['Authorization'] = options.authorization
        }
        return (await this.executor({uri: _uri, method: 'PATCH', headers: _headers, body: options.body})) as Promise<LoginTransferUpdateResponse>;
    }
}

export type LoginTransferControllerOptions = {
    'create': {}, 
    'get': {
        readonly id: string, 
        readonly authorization?: string | undefined
    }, 
    'claim': {
        readonly body: LoginTransferClaimRequest
    }, 
    'update': {
        readonly id: string, 
        readonly body: LoginTransferUpdateRequest, 
        readonly authorization?: string | undefined
    }, 
    'cancel': {
        readonly id: string
    }, 
    'createToken': {
        readonly id: string, 
        readonly authorization?: string | undefined
    }
}

File: server/src/main/kotlin/com/coooolfan/unirhy/controller/LoginTransferController.kt
package com.coooolfan.unirhy.controller

import cn.dev33.satoken.annotation.SaCheckLogin
import cn.dev33.satoken.stp.StpUtil
import com.coooolfan.unirhy.error.CommonException
import com.coooolfan.unirhy.error.LoginTransferException
import com.coooolfan.unirhy.model.LoginTransfer
import com.coooolfan.unirhy.model.by
import com.coooolfan.unirhy.model.dto.LoginTransferCreateResponse
import com.coooolfan.unirhy.model.dto.LoginTransferUpdateRequest
import com.coooolfan.unirhy.model.dto.LoginTransferUpdateResponse
import com.coooolfan.unirhy.service.LoginTransferAuthorization
import com.coooolfan.unirhy.service.LoginTransferService
import jakarta.servlet.http.HttpServletResponse
import org.babyfish.jimmer.client.FetchBy
import org.babyfish.jimmer.sql.fetcher.Fetcher
import org.babyfish.jimmer.sql.kt.fetcher.newFetcher
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PatchMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestHeader
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestController
import java.util.UUID

@RestController
@RequestMapping("/api/login-transfers")
class LoginTransferController(
    private val service: LoginTransferService,
) {
    @SaCheckLogin
    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    fun create(response: HttpServletResponse): LoginTransferCreateResponse {
        val result = service.create(StpUtil.getLoginIdAsLong())
        response.setHeader(HttpHeaders.LOCATION, "/api/login-transfers/${result.id}")
        response.setHeader(HttpHeaders.CACHE_CONTROL, NO_STORE)
        return result
    }

    /**
     * 查询登录交接状态
     *
     * 原设备凭账号会话查询，新设备凭认领访问令牌查询，后者只获得完成流程所需的最小状态。
     * 已进入终态（含过期）的交接同样返回 `200`，由调用方读取 `status` 判断。
     */
    @GetMapping("/{id}")
    @Throws(
        CommonException.AuthenticationFailed::class,
        CommonException.Forbidden::class,
        LoginTransferException.NotFound::class,
    )
    fun get(
        @PathVariable id: UUID,
        @RequestHeader(HttpHeaders.AUTHORIZATION, required = false) authorization: String?,
        response: HttpServletResponse,
    ): @FetchBy("SOURCE_TRANSFER_FETCHER") LoginTransfer {
        response.setHeader(HttpHeaders.CACHE_CONTROL, NO_STORE)
        return when (val auth = resolveAuthorization(authorization)) {
            is LoginTransferAuthorization.Account ->
                service.getForAccount(id, auth.accountId, SOURCE_TRANSFER_FETCHER)

            is LoginTransferAuthorization.Claim ->
                service.getForClaim(id, auth.token, CLAIM_TRANSFER_FETCHER)

            LoginTransferAuthorization.None -> throw CommonException.AuthenticationFailed()
        }
    }

    @PatchMapping("/{id}")
    @Throws(
        CommonException.AuthenticationFailed::class,
        CommonException.Forbidden::class,
        CommonException.InvalidRequest::class,
        LoginTransferException.NotFound::class,
        LoginTransferException.StatusConflict::class,
        LoginTransferException.Expired::class,
    )
    fun update(
        @PathVariable id: UUID,
        @RequestBody request: LoginTransferUpdateRequest,
        @RequestHeader(HttpHeaders.AUTHORIZATION, required = false) authorization: String?,
        response: HttpServletResponse,
    ): LoginTransferUpdateResponse {
        response.setHeader(HttpHeaders.CACHE_CONTROL, NO_STORE)
        return service.applyUpdate(
            id = id,
            auth = resolveAuthorization(authorization),
            request = request,
            sourceFetcher = SOURCE_TRANSFER_FETCHER,
            claimFetcher = CLAIM_TRANSFER_FETCHER,
        )
    }

    @SaCheckLogin
    @DeleteMapping("/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun cancel(@PathVariable id: UUID) {
        service.cancel(id, StpUtil.getLoginIdAsLong())
    }

    @PostMapping("/{id}/tokens")
    @ResponseStatus(HttpStatus.CREATED)
    fun createToken(
        @PathVariable id: UUID,
        @RequestHeader(HttpHeaders.AUTHORIZATION, required = false) authorization: String?,
        response: HttpServletResponse,
    ): TokenLoginResponse {
        val auth = resolveAuthorization(authorization) as? LoginTransferAuthorization.Claim
            ?: throw CommonException.AuthenticationFailed()
        val token = service.createToken(id, auth.token)
        response.setHeader(HttpHeaders.LOCATION, "/api/tokens/current")
        response.setHeader(HttpHeaders.CACHE_CONTROL, NO_STORE)
        return TokenLoginResponse(token)
    }

    private fun resolveAuthorization(authorization: String?): LoginTransferAuthorization {
        val accountLoggedIn = StpUtil.isLogin()
        val claimToken = authorization
            ?.takeIf { it.startsWith(BEARER_PREFIX, ignoreCase = true) }
            ?.substring(BEARER_PREFIX.length)
            ?.trim()
            ?.takeIf { it.isNotEmpty() }
        if (authorization != null && claimToken == null) throw CommonException.AuthenticationFailed()
        if (accountLoggedIn && claimToken != null) throw CommonException.InvalidRequest()
        return when {
            accountLoggedIn -> LoginTransferAuthorization.Account(StpUtil.getLoginIdAsLong())
            claimToken != null -> LoginTransferAuthorization.Claim(claimToken)
            else -> LoginTransferAuthorization.None
        }
    }

    companion object {
        /** 原设备视图：可以看到认领设备的信息。 */
        val SOURCE_TRANSFER_FETCHER: Fetcher<LoginTransfer> =
            newFetcher(LoginTransfer::class).by {
                deviceName()
                platform()
                clientVersion()
                status()
                createdAt()
                expiresAt()
                claimedAt()
                authorizedAt()
                closedAt()
            }

        /**
         * 新设备视图：只包含推进流程所需的状态与时间。
         *
         * 这里刻意不返回账号或设备信息——今后新增的原设备专属字段，
         * 只有显式加进本 fetcher 才会对新设备可见。
         */
        val CLAIM_TRANSFER_FETCHER: Fetcher<LoginTransfer> =
            newFetcher(LoginTransfer::class).by {
                status()
                createdAt()
                expiresAt()
                authorizedAt()
                closedAt()
            }

        private const val BEARER_PREFIX = "Bearer "
        private const val NO_STORE = "no-store"
    }
}
